planner/core/rule_partition_processor.go

```diff
@@ -190,14 +190,24 @@ func (s *partitionProcessor) findUsedPartitions(ctx sessionctx.Context, tbl tabl
 						}
 						used = append(used, int(idx))
 					}
-				} else {
-					used = []int{FullRange}
-					break
+					continue
+				}
+
+				// maximum number of partitions is 1024
+				if col.RetType.Flen < 10 {
+					// all possible hash values
+					maxUsedPartitions := 1 << col.RetType.Flen
+					if maxUsedPartitions < numPartitions {
+						for i := 0; i < maxUsedPartitions; i++ {
+							used = append(used, i)
```

Comment:
Should we handle `null` specially here?

Reply:
Now, the new commit should ensure the `used` is not null.